[ADD] sale line conf price total test

mourad-ehm opened the pull request.

mourad-ehm committed:
[ADD] sale line conf price total test

sebastienbeau commented on sale_configurator_option/tests/test_sale_order.py:
```diff
@@ -11,8 +11,34 @@ class SaleOrderCase(SavepointCase):
         self.assertEqual(self.sale.amount_tax, 16.5)
+
+    def test_conf_total_amount(self):
+        self.assertEqual(self.line_with_opt.price_subtotal, 0)
```
no necessary

sebastienbeau commented on sale_configurator_option/tests/test_sale_order.py:
```diff
@@ -11,8 +11,34 @@ class SaleOrderCase(SavepointCase):
+        self.assertEqual(self.line_with_opt.price_config_subtotal, 110)
+        self.assertEqual(self.line_with_opt.price_config_total, 126.5)
+        # Set price to product with option and check totals
+        cls.line_with_opt.price_unit = 100
```
move in a new test

mourad-ehm replied:
ping @sebastienbeau

mourad-ehm committed:
[FIX] sale line conf price total test

mourad-ehm committed:
[ADD] sale line conf price total compute

mourad-ehm committed:
[FIX] sale line conf price total test

mourad-ehm committed:
[ADD] product configurator option

sebastienbeau commented on sale_configurator_option/models/product.py:
```diff
@@ -0,0 +1,76 @@
+    product_tmpl_id = fields.Many2one(
+        'product.template', 'Product Template',
+        auto_join=True, index=True, ondelete="cascade", required=True)
+    product_config_id = fields.Many2one(
```
product_option_id

sebastienbeau replied:
or option_id

sebastienbeau replied:
product_id so we have the same datamodel as mrp.bom.line

mourad-ehm committed:
[FIX] del opt_qty

sebastienbeau commented on sale_configurator_option/models/product.py:
```diff
@@ -0,0 +1,76 @@
+    sequence = fields.Integer(
+        'Sequence', default=1,
+        help="Gives the sequence order when displaying.")
+    opt_qty = fields.Float(
```
useless?

sebastienbeau commented on sale_configurator_option/models/product.py:
```diff
@@ -0,0 +1,76 @@
+class ProductTemplate(models.Model):
+    _inherit = 'product.template'
+
+    configurable_opt_ok = fields.Boolean(
```
is_configurable

mourad-ehm committed:
[IMP] add config product onchange code & test - rename product_config_id to product_id

mourad-ehm committed:
[FIX] flake8

sebastienbeau commented on sale_configurator_option/tests/test_sale_order.py:
```diff
@@ -40,3 +52,23 @@ def test_conf_total_amount_main_without_price(self):
+        self.line_with_opt_2.product_id_change()
+        default_options = [
+            x.product_id
+            for x in self.product_with_option_2.product_tmpl_id.\
```
you can use

self.product_with_option_2.product_config_opt_ids

sebastienbeau commented on sale_configurator_option/models/product.py:
```diff
@@ -67,9 +67,9 @@ def onchange_product_uom_id(self):
+        'Is a Configurable Product ?',
+        help='Chek this, if the product is configurable with options',
     )
     product_config_opt_ids = fields.One2many(
```
configurable_option_ids

sebastienbeau replied:
configurable_option_ids

sebastienbeau commented on sale_configurator_option/tests/test_sale_order.py:
```diff
@@ -40,3 +52,23 @@ def test_conf_total_amount_main_without_price(self):
+            for x in self.product_with_option_2.product_tmpl_id.\
+                    product_config_opt_ids
+            if x.opt_default_qty > 0]
+        self.assertEqual(len(self.line_with_opt_2.option_ids), len(default_options))
```
Something like that will be better
product_ids = set(line.mapped("option_ids.product_id.id"))
self.assertEqual(product_ids, set([self.product_option_1.id, self.product_option_2.id])

sebastienbeau replied:
self.assertEqual(len(line.option_ids), 2)

mourad-ehm committed:
[FIX] Test

mourad-ehm committed:
[IMP] Add sale line configurator view

mourad-ehm committed:
[IMP] Add option qty management

sebastienbeau commented on sale_configurator_option/models/product.py:
```diff
@@ -0,0 +1,76 @@
+        auto_join=True, index=True, ondelete="cascade", required=True)
+    product_id = fields.Many2one(
+        'product.product', 'Option', required=True)
+    product_uom = fields.Many2one(
```
please keep product_uom_id
In order to be consistente with mrp.bom.line : https://github.com/odoo/odoo/blob/13.0/addons/mrp/models/mrp_bom.py#L265

mourad-ehm committed:
[Add] option qty type

mourad-ehm committed:
[Add] sale_configurator_variant

sebastienbeau commented on sale_configurator_variant/models/sale.py:
```diff
@@ -0,0 +1,116 @@
+    @api.depends("variant_ids.product_uom_qty")
+    def _compute_is_variant_qty_need_compute(self):
+        for record in self:
+            record.is_variant_qty_need_compute = True
```
Can you check the qty and only set true if !=

sebastienbeau replied:
```
def _get_parent_qty(self):
    self.ensure_one()
    return sum(self.variant_ids.mapped("product_uom_qty")

def _compute_is_variant_qty_need_compute(self):
    for record in self:
        qty = record._get_parent_qty()
        record.is_variant_qty_need_recompute = qty != record.product_uom_qty

```

sebastienbeau commented on sale_configurator_variant/models/sale.py:
```diff
@@ -0,0 +1,116 @@
+        change_default=True,
+        ondelete="restrict",
+    )
+    is_variant_qty_need_compute = fields.Boolean(
```
need_recompute

sebastienbeau commented on sale_configurator_variant/models/sale.py:
```diff
@@ -0,0 +1,116 @@
+            record.is_variant_qty_need_compute = True
+
+    @api.multi
+    def _compute_parent_variant_qty(self):
```
_set_parent_variant_qty

mourad-ehm committed:
[Add] refactore code & integrate review code

mourad-ehm committed:
[IMP] price multivariant management & fix bugs & add test

mourad-ehm commented:
Hi @sebastienbeau could you review code ?

sebastienbeau commented on sale_configurator_option/models/sale.py:
```diff
@@ -1,13 +1,110 @@
+        default=1.0,
+    )
+    parent_option_qty = fields.Float(related="parent_option_id.product_uom_qty",)
+    force_option_qty = fields.Boolean("Force Option Qty",)
```
not needed

sebastienbeau commented on sale_configurator_variant/models/sale.py:
```diff
@@ -0,0 +1,135 @@
+        for record in records:
+            qty = record._get_child_qty()
+            record.is_variant_qty_need_recompute = qty != record.product_uom_qty
+            record._set_parent_variant_qty()
```
not needed

sebastienbeau commented on sale_configurator_variant/models/sale.py:
```diff
@@ -0,0 +1,135 @@
+            for variant in self.product_tmpl_id.product_variant_ids:
+                variant_lines.append((0, 0, self._prepare_sale_line_variant(variant)))
+            self.variant_ids = variant_lines
+        if self.product_tmpl_id:
```
TODO we need to check how to fix this

mourad-ehm committed:
[FIX] Clean code

mourad-ehm committed:
[FIX] add product configurator option on sale line id

mourad-ehm committed:
[ADD] module sale_configurator_option_typology

mourad-ehm committed:
[FIX] sale_configurator_option_typology add view & demo & security data

mourad-ehm commented:
hi @sebastienbeau I added module sale_configurator_option_typology. Could you review please ?

mourad-ehm committed:
[FIX] add is_option flag to product and sql constarint

mourad-ehm committed:
[ADD] module sale_configurator_option_link

mourad-ehm committed:
[FIX] bugs & add _set_included_options

mourad-ehm committed:
[FIX] Flak8

sebastienbeau commented on sale_configurator_base/models/sale.py:
```diff
@@ -0,0 +1,76 @@
+    pricelist_id = fields.Many2one(
+        related="order_id.pricelist_id", string="Pricelist", store=True, readonly=True
+    )
+    is_configurable_parent_opt = fields.Boolean(
```
Having a computed field seem to be a better idea.
It will avoid to put too many logic in onchange
Also I see that this option is only related to option, maybe there is something wrong

sebastienbeau commented on sale_configurator_option/models/product.py:
```diff
@@ -0,0 +1,116 @@
+        (
+            "product_tmpl_id_product_id_unique",
+            "UNIQUE(product_tmpl_id,product_id)",
+            "Option must be unique by configable product",
```
```suggestion
            "Option must be unique by configurable product",
```

sebastienbeau commented on sale_configurator_option/tests/test_sale_order.py:
```diff
@@ -11,8 +11,76 @@ class SaleOrderCase(SavepointCase):
+        self.assertEqual(self.line_opt_2.price_config_total, 0)
+        self.assertEqual(self.line_opt_3.price_config_total, 0)
+
+    def test_conf_total_amount_main_without_price(self):
```
Wrong method name
```suggestion
    def test_conf_total_amount_main_with_price(self):
```

sebastienbeau commented on sale_configurator_option/tests/test_sale_order.py:
```diff
@@ -11,8 +11,76 @@ class SaleOrderCase(SavepointCase):
+        opt_1 = self.env.ref("sale_configurator_option.product_configurator_option_1")
+        opt_2 = self.env.ref("sale_configurator_option.product_configurator_option_2")
+
+        qties = {
```
It always better to avoid logic in test.
Having something like
options = new_line.option_ids
self.assertEqual(options[0].product_uom_qty, 1)
self.assertEqual(options[2].product_uom_qty, 3)
is more simple to read and more explicit on the result expected

sebastienbeau commented on sale_configurator_option_link/models/product.py:
```diff
@@ -0,0 +1,45 @@
+
+    included_by_option_id = fields.Many2one(
+        "product.configurator.option",
+        "Inclueded by Option",
```
```suggestion
        "Included by Option",
```

mourad-ehm committed:
[FIX] included_by_product constraint & compute

sebastienbeau committed:
[IMP] in case that we create the line by passing the data in option_ids addthe missing order_id

sebastienbeau committed:
[IMP] add a test that break the code

sebastienbeau commented on sale_configurator_option/models/sale.py:
```diff
@@ -1,13 +1,126 @@
+            self.product_option_id = option_id
+        return res
+
+    @api.onchange("product_uom", "option_unit_qty", "product_uom_qty")
```
we should do this using a computed field "is_option_qty_need_recompute" and then post processing the with
     def _recompute_done(self, field):

mourad-ehm committed:
Apply suggestions from code review

Co-authored-by: beau sebastien <sebastien.beau@akretion.com>

mourad-ehm commented on sale_configurator_variant/models/sale.py:
```diff
@@ -0,0 +1,128 @@
+        return records
+
+    def _recompute_done(self, field):
+        super(SaleOrderLine, self)._recompute_done(field)
```
```suggestion
        super()._recompute_done(field)
```

mourad-ehm commented on sale_configurator_variant/models/sale.py:
```diff
@@ -0,0 +1,128 @@
+        """
+        get the config subtotal amounts of the SO line.
+        """
+        res = super(SaleOrderLine, self)._get_price_config_subtotal()
```
```suggestion
        res = super()._get_price_config_subtotal()
```

mourad-ehm commented on sale_configurator_variant/models/sale.py:
```diff
@@ -0,0 +1,128 @@
+        """
+        get the config subtotal amounts of the SO line.
+        """
+        res = super(SaleOrderLine, self)._get_price_config_total()
```
```suggestion
        res = super()._get_price_config_total()
```

mourad-ehm commented on sale_configurator_option/models/sale.py:
```diff
@@ -1,13 +1,126 @@
+        """
+        get the config subtotal amounts of the SO line.
+        """
+        res = super(SaleOrderLine, self)._get_price_config_subtotal()
```
```suggestion
        res = super()._get_price_config_subtotal()
```

mourad-ehm commented on sale_configurator_option/models/sale.py:
```diff
@@ -1,13 +1,126 @@
+        """
+        get the config subtotal amounts of the SO line.
+        """
+        res = super(SaleOrderLine, self)._get_price_config_total()
```
```suggestion
        res = super()._get_price_config_total()
```

mourad-ehm commented on sale_configurator_option/models/sale.py:
```diff
@@ -1,13 +1,126 @@
+
+    @api.onchange("product_id")
+    def product_id_change(self):
+        res = super(SaleOrderLine, self).product_id_change()
```
```suggestion
        res = super().product_id_change()
```

mourad-ehm commented on sale_configurator_option/models/sale.py:
```diff
@@ -1,13 +1,126 @@
+                    opt.product_uom_qty = opt.option_unit_qty * self.product_uom_qty
+                    opt.product_uom_change()
+
+        res = super(SaleOrderLine, self).product_uom_change()
```
```suggestion
        res = super().product_uom_change()
```

mourad-ehm commented on sale_configurator_option_typology/models/sale.py:
```diff
@@ -0,0 +1,25 @@
+
+    @api.onchange("product_option_id")
+    def product_option_id_change(self):
+        res = super(SaleOrderLine, self).product_option_id_change()
```
```suggestion
        res = super().product_option_id_change()
```

mourad-ehm committed:
Remove classies from super

mourad-ehm commented on sale_configurator_base/models/sale.py:
```diff
@@ -0,0 +1,76 @@
+# @author Sébastien BEAU <sebastien.beau@akretion.com>
+# @author Mourad EL HADJ MIMOUNE <mourad.elhadj.mimoune@akretion.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
```
```suggestion
```

mourad-ehm commented on sale_configurator_option_link/models/sale.py:
```diff
@@ -0,0 +1,29 @@
+# @author Sébastien BEAU <sebastien.beau@akretion.com>
+# @author Mourad EL HADJ MIMOUNE <mourad.elhadj.mimoune@akretion.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
```
```suggestion
```

mourad-ehm commented on sale_configurator_option_typology/models/sale.py:
```diff
@@ -0,0 +1,25 @@
+# @author Sébastien BEAU <sebastien.beau@akretion.com>
+# @author Mourad EL HADJ MIMOUNE <mourad.elhadj.mimoune@akretion.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
```
```suggestion
```

mourad-ehm commented on sale_configurator_variant/models/sale.py:
```diff
@@ -0,0 +1,128 @@
+# @author Sébastien BEAU <sebastien.beau@akretion.com>
+# @author Mourad EL HADJ MIMOUNE <mourad.elhadj.mimoune@akretion.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
```
```suggestion
```

mourad-ehm committed:
Delete useless blank lines

mourad-ehm committed:
[FIX]  api.model_create_multi & add is_option_qty_need_recompute

mourad-ehm committed:
Apply new super synthax

mourad-ehm committed:
[IMP] add product_configurator_template

mourad-ehm committed:
[FIX] add product_configurator_template view & fix access rulle & onchange option

mourad-ehm committed:
[FIX] Filter related option only on sale line & pep8

mourad-ehm committed:
[ADD] sale_configurator_option_restricted_qty module

mourad-ehm committed:
[FIX] restricted_qty sale view

sebastienbeau commented:
@mourad-ehm tu pourras regarder pour corriger travis? Merci

sebastienbeau committed:
[FIX] fix black

sebastienbeau committed:
[FIX] fix qty on option when creating sale order

sebastienbeau committed:
[IMP] avoid storing pricelist as it's impact write and we do not use it, also remove useless readonly as it's the default

sebastienbeau committed:
Merge branch '12.0' into 12.0_add_sale_conf_price_test

sebastienbeau committed:
[FIX] fix travis

sebastienbeau committed:
[REF] rename default_code into code

sebastienbeau committed:
[IMP] improve menu and add glue module with pim

sebastienbeau committed:
[IMP] improve UI of configurator template

sebastienbeau commented:
TODO
- [x] fix test
- [x] fix broken UI (button edit save when adding a new option...)
- [ ] fix sorting line that mix everything
- [ ] issue with min/max qty (the min max qty can be for the qty total or the qty or operation per product)

sebastienbeau committed:
[IMP] start to clean UI

sebastienbeau committed:
[IMP] improve pop-up configurator ui

sebastienbeau committed:
[FIX] fix visibility

sebastienbeau committed:
[REF] simplify and fix view

Use a dedicated view for option to duplicated xml code

sebastienbeau committed:
[FIX] fix some warning

sebastienbeau committed:
[REF] rename sale_configurator_option_typology to sale_configurator_option_area and change the logic

sebastienbeau committed:
[IMP] add setup

sebastienbeau committed:
fixup! [REF] rename sale_configurator_option_typology to sale_configurator_option_area and change the logic

sebastienbeau committed:
[FIX] fix uniq constraint

sebastienbeau committed:
[IMP] improve UX for area and refactor code

sebastienbeau committed:
[FIX] fix dependency

mourad-ehm committed:
[FIX] bug product option without conf & decoration-info for sale line with option & manage default option

mourad-ehm committed:
[IMP] improuve product multi variant Ui

mourad-ehm committed:
[FIX] stash error

mourad-ehm committed:
[FIX] sale_max_qty default=0

sebastienbeau committed:
[CLEAN] run pre-commit

sebastienbeau committed:
[REF] fix test and add missing scenario with fix and refactor

- refactor is_configurable_parent_opt
   - is_configurable is now a computed field in based module
   - is_configurable_opt is a related field
- fix computed field (missing depends)
- add tests

sebastienbeau committed:
[REF] fix test and add missing scenario with fix and refactor

- refactor is_configurable_parent_opt
   - is_configurable is now a computed field in based module
   - is_configurable_opt is a related field
- fix computed field (missing depends)
- add test

sebastienbeau commented:
@mourad-ehm last thing before merging is make readonly the option line in the tree view

sebastienbeau commented:
We should also fix the reordering

mourad-ehm committed:
Merge branch '12.0_add_sale_conf_price_test' of https://github.com/akretion/sale-configurator into 12.0_add_sale_conf_price_test

mourad-ehm committed:
[FIX] add manual_sale min max qty on product.configurator.option

mourad-ehm committed:
[FIX] Configutaor views (add fields to be set correctely by onchange)

mourad-ehm committed:
[FIX] add redme files and pre-commit format

mourad-ehm committed:
[FIX]  sale_configurator_option_area DESCRIPTION.rst

sebastienbeau committed:
[IMP] Update precommit config

sebastienbeau committed:
[FIX] fix precommit

sebastienbeau committed:
[FIX] fix precommit

mourad-ehm committed:
[FIX] remove sequence handle from sale_configurator_option sale_configurator_variant to fix dysply bug

sebastienbeau committed:
[FIX] fix precommit

sebastienbeau commented:
TODO
- [ ] support correctly the sequence instead of removing it (modifying an existing line will add the option at the end)

mourad-ehm committed:
[FIX] sale_configurator_option_link view

mourad-ehm committed:
[FIX] pre commi format

mourad-ehm committed:
[FIX] sale_configurator_variant tests

sebastienbeau committed:
[IMP] keep sale order line ordered correctly

sebastienbeau committed:
[IMP] change button position

mourad-ehm committed:
[FIX] sale_configurator_option delete option default_qty

mourad-ehm committed:
[FIX] remove module imported but unused

mourad-ehm committed:
[FIX] test
